File: day02.py
#!/usr/bin/env pipenv run python
"""Solutions to day 2 of Advent of Code"""

from collections import defaultdict, Counter
from get_input import get_input

def part1(lines):
    """Solution to part 1"""
    checksum = defaultdict(int)
    for line in lines:
        digits = Counter(list(line))
        for i in range(2, 5):
            if i in digits.values():
                checksum[i] += 1
    result = 1
    for num in checksum.values():
        result *= num
    return result


def part2(lines):
    """Solution to part 2"""
    for h_index, head in enumerate(lines):
        for tail in lines[:h_index]:
            common_letters = ''.join(
                h for h, t in zip(list(head), list(tail)) if h == t)
            if len(common_letters) + 1 == len(head):
                return common_letters
    raise Exception("Could not find soltution")

if __name__ == '__main__':
    LINES = get_input(day=2, year=2018).splitlines()
    print("Part 1: {}".format(part1(LINES)))
    print("Part 2: {}".format(part2(LINES)))

File: day02.rs
extern crate common;

use std::collections::HashMap;

use common::get_input;

fn part1(lines: &Vec<String>) -> u32 {
    let mut threes = 0;
    let mut twos = 0;
    for line in lines {
        let mut counts:HashMap<char, u32> = HashMap::new();
        line.chars().for_each(|c| *counts.entry(c).or_insert(0) += 1);
        if counts.values().any(|&v| v == 3) { threes += 1 }
        if counts.values().any(|&v| v == 2) { twos += 1 }
    }
    return threes * twos;
}

fn part2(lines: &Vec<String>) -> String {
    for (l, head) in lines.into_iter().enumerate() {
        for tail in lines.into_iter().skip(l) {
            let same:String = head
                .chars()
                .into_iter()
                .zip(tail.chars())
                .filter_map(
                    |(a, b)| if a == b { Some(a) } else { None })
                .collect();
            if head.chars().count() - 1 == same.chars().count() {
                return same
            }
        }
    }
    panic!("Cannot find solution");
}

fn parse(lines: String) -> Vec<String> {
    return lines
        .trim()
        .split("\n")
        .into_iter()
        .map(|s| s.to_string())
        .collect();
}

fn main() {
    let input = parse(get_input(02, 2018));
    print!("Part 1: {}\n", part1(&input));
    print!("Part 2: {}\n", part2(&input));
}
